### PR TITLE
feat: added sourcify with tests

### PR DESCRIPTION
**Pull request by 0xvangrim**
# Description

Issue: https://github.com/DXgovernance/dxvote/issues/755

Adding a Sourcify API call to see if the contract chosen is verified or not. If it is chosen, we will decode the ABI and give the user an interface.

image

Difficult to test, but I added tests and tested it so it doesn't break any of the existing logic we use to show the contract interface.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Manually and with unit tests.

# Checklist:

- [X ] My code follows the existing style of this project
- [X ] I have performed a self-review of my own code
- [ X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ X] I have added tests that prove my fix is effective or that my feature works
- [X ] New and existing unit tests pass locally with my changes
- [ ] Any UI changes have been tested and made responsive for mobile views

### CODE REVIEW

File: src/hooks/Guilds/contracts/useDecodedCall.ts [hunk 1]
```diff
@@ -124,21 +131,39 @@ export const bulkDecodeCallsFromOptions = (
...
+      );
+    } else {
+      setDecodedCall(null);
+    }
+    return () => {
+      isCancelled.current = true;
+    };
+  }, [call, contracts, chainId]);
+
+  return decodedCall || { decodedCall: null, contract: null };
```
Comment (MiltonTulli) [inline src/hooks/Guilds/contracts/useDecodedCall.ts]: missing approval in the fallback object.

File: src/utils/sourcify.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+export const lookUpContractWithSourcify = async ({ chainId, address }) => {
+  const baseUrl = `https://sourcify.dev/server/files/any`;
```
Comment (MiltonTulli) [inline src/utils/sourcify.ts:2]: Do you think we store this in env/configs/constants file?

**Review by MiltonTulli — DISMISSED** [review]

File: src/utils/sourcify.ts (hunk 2 above)
Comment (0xvangrim) [reply src/utils/sourcify.ts:2]: Sourcify will be used to lookup contracts that we are not storing ourselves. Or am I missing something in your question?

Comment (MiltonTulli) [reply src/utils/sourcify.ts:2]: Just wondering if we should pull this url from config file, or env. Nothing serious

**Review by MiltonTulli — APPROVED** [review]

**Review by AugustoL — CHANGES_REQUESTED** [review]
The only thing that It is missing is a test comparing a the abi and contact byecode hash from the richContractData with one from sourcify, I see you are only testing against a random contract/

**Comment by 0xvangrim** [comment]
> The only thing that It is missing is a test comparing a the abi and contact byecode hash from the richContractData with one from sourcify, I see you are only testing against a random contract/

Hm, I'm not certain why we would need to have that test? The `richContractData` and `sourcify` should never overlap. When the contract is not in the contract registry we will fetch it from sourcify, but only then. We will never make a double-check against both the registry and sourcify.

File: src/utils/sourcify.ts (hunk 2 above)
Comment (0xvangrim) [reply src/utils/sourcify.ts:2]: Ah I see, now I understand!

**Comment by 0xvangrim** [comment]
@AugustoL I've tested some of our `richContractData` that exists in our `dxvote`-file now and they haven't been been verified there yet.

**Review by AugustoL — APPROVED** [review]

**Comment by AugustoL** [comment]
Approved, we see how we can add those tests for sourcify after then

**Pull request merged**